File: common/rtl/src/main/scala/zoo/common/architecture/Operations.scala
package zoo.common.architecture

/**
 * Base trait for operation execution capabilities.
 */
trait Operations

/**
 * Fixed-point (integer) and logical operations.
 * Found in almost all machines.
 */
trait FixedPointOperations extends Operations {
  val hasMultiplyDivide: Boolean
}

/**
 * Floating-point (real) operations.
 * Found in scientific processors.
 * Example: IBM System/360 Model 91, Cray-1.
 */
trait FloatingPointOperations extends Operations {
  val fpuPrecision: Int // E.g., 32, 64, or 80 bits
}

/**
 * Decimal and character string operations.
 * Found in business-oriented processors.
 * Example: IBM 1401, IBM System/360 (decimal package).
 */
trait DecimalOperations extends Operations {
  val hasDecimalPack: Boolean
}

/**
 * Vector execution operations.
 * Operations executed concurrently/pipelined over an array of data.
 * Example: Cray-1.
 */
trait VectorOperations extends Operations {
  val hasChaining: Boolean // Cray-1 style functional unit chaining
}

File: common/sw/architecture/operations.py
from abc import ABC

class Operations(ABC):
    """
    Base class for operation execution capabilities.
    """
    pass

class FixedPointOperations(Operations):
    """
    Fixed-point (integer) and logical operations.
    """
    has_multiply_divide: bool

class FloatingPointOperations(Operations):
    """
    Floating-point (real) operations.
    """
    fpu_precision: int

class DecimalOperations(Operations):
    """
    Decimal and character string operations.
    """
    has_decimal_pack: bool

class VectorOperations(Operations):
    """
    Vector execution operations.
    """
    has_chaining: bool
